src/sage/graphs/graph.py

```diff
@@ -4328,6 +4328,228 @@ def weight(x):
+            Traceback (most recent call last):
+            ...
+            ValueError: the input is not a matching
+            sage: G = graphs.CompleteBipartiteGraph(3, 3)
+            sage: M = [(2*i, 2*i + 1) for i in range(4)]
+            sage: u = 0
+            sage: S = G.M_alternating_even_mark(u, M)
+            Traceback (most recent call last):
+            ...
+            ValueError: the input is not a matching of the graph
+
+        REFERENCES:
+
+        - [LR2004]_
+
+        .. SEEALSO::
+            :meth:`~sage.graphs.graph.Graph.is_factor_critical`,
+            :meth:`~sage.graphs.graph.Graph.is_matching_covered`,
+            :meth:`~sage.graphs.graph.Graph.is_bicritical`
+        """
+
+        G = self.to_simple()
+
+        # The input vertex must be a valid vertex of the graph
+        if vertex not in G.vertices():
+            raise ValueError("'{}' is not a vertex of the graph".format(vertex))
+
+        # The input matching must be a valid matching of the graph
+        M = Graph(matching)
+        if any(d != 1 for d in M.degree()):
+            raise ValueError("the input is not a matching")
+        if not M.is_subgraph(G, induced=False):
+            raise ValueError("the input is not a matching of the graph")
+
+        # Build an M-alternating tree T rooted at vertex
+        import itertools
+        from queue import Queue
+
+        q = Queue()
+        q.put(vertex)
+
+        even = set([vertex])
+        odd = set()
+        predecessor = {vertex: vertex}
+        rank = {vertex: 0}
+
+        if vertex in M.vertices():
+            u = M.neighbors(vertex)[0]
+            predecessor[u] = None
+            rank[u] = -1
+            odd.add(u)
+
+        while not q.empty():
+            x = q.get()
+            for y in G.neighbor_iterator(x):
+                if y in odd:
+                    continue
+                elif y in even:
+                    # Search t := LCA(x, y)
+                    ancestor_x = [x]
+                    ancestor_y = [y]
+
+                    while ancestor_x[-1] != ancestor_y[-1]:
+                        if rank[ancestor_x[-1]] > rank[ancestor_y[-1]]:
+                            ancestor_x.append(predecessor[ancestor_x[-1]])
+                        elif rank[ancestor_x[-1]] < rank[ancestor_y[-1]]:
+                            ancestor_y.append(predecessor[ancestor_y[-1]])
+                        else:
+                            ancestor_x.append(predecessor[ancestor_x[-1]])
+                            ancestor_y.append(predecessor[ancestor_y[-1]])
+
+                    lcs = ancestor_x.pop()
+                    ancestor_y.pop()
+                    # Set t as pred of all vertices of the chains and add
+                    # vertices marked odd to the queue
+                    for a in itertools.chain(ancestor_x, ancestor_y):
+                        predecessor[a] = lcs
+                        rank[a] = rank[lcs] + 1
```

Comment:
you may put the value `rank[lcs] + 1` in a variable to avoid recomputing this value